### PR TITLE
Support --hostnames

### PR DESCRIPTION
**Pull request by alexey-pelykh**

### CODE REVIEW

File: proxy/proxy.py [hunk 1]
```diff
@@ -211,16 +211,18 @@ def setup(self) -> None:
...
-            ports.append(
+            ports.add(
                 cast(
                     'TcpSocketListener',
                     self.listeners.pool[index],
                 )._port,
             )
-        self.flags.ports = ports
+        if self.flags.port in ports:
+            ports.remove(self.flags.port)
```
Comment (abhinavsingh) [inline proxy/proxy.py:224]: Nice catch here 👍

File: Makefile [hunk 2]
```diff
@@ -161,7 +162,7 @@ lib-speedscope:
 		-o profile.speedscope.json \
 		-f speedscope \
 		-t -F -s -- \
-		python -m proxy \
+		$(PYTHON) -m proxy \
```
Comment (abhinavsingh) [inline Makefile:165]: Thanks for getting these addressed too

File: README.md [hunk 3]
```diff
@@ -213,7 +213,8 @@
     - `--enable-reverse-proxy --plugins proxy.plugin.ReverseProxyPlugin`
   - Plugin API is currently in *development phase*. Expect breaking changes. See [Deploying proxy.py in production](#deploying-proxypy-in-production) on how to ensure reliability across code changes.
 
-- Can listen on multiple ports
+- Can listen on multiple addresses and ports
+  - Use `--hostnames` flag to provide additional addresses
```
Comment (abhinavsingh) [inline README.md:217]: +1 for the doc update :)

File: proxy/common/flag.py [hunk 4]
```diff
@@ -301,14 +307,7 @@ def initialize(
...
             )
         else:
-            # FIXME: Not true for tests, as this value will be a mock.
-            #
-            # It's a problem only on Windows.  Instead of a proper
-            # fix in the tests, simply commenting this line of assertion
-            # for now.
-            #
-            # assert args.unix_socket_path is None
-            args.family = socket.AF_INET6 if args.hostname.version == 6 else socket.AF_INET
```
Comment (abhinavsingh) [inline proxy/common/flag.py:311]: Wondering if this is intentional.  Can you provide some background.  Thank you

PS: I see all windows workflows are kind of broken now , e.g. https://github.com/abhinavsingh/proxy.py/actions/runs/4533948198/jobs/7987500924?pr=1325

**Review by abhinavsingh — COMMENTED** [review]
LGTM.  PTAL at my comment related to windows & its workflow.

Thank you

File: proxy/common/flag.py (hunk 4 above)
Comment (alexey-pelykh) [reply proxy/common/flag.py:311]: Looking into that issue right now as that code can't be valid with mixed IPv4/IPv6 addresses.

Comment (alexey-pelykh) [reply proxy/common/flag.py:311]: Most tests mocking `socket.fromfd` are quite hard to redo w/o issues tbh

Comment (abhinavsingh) [reply proxy/common/flag.py:311]: > assert args.unix_socket_path is None

The comment above this line indicates issue on windows after removing this line.  I'll need to get my hands dirty to recall more context around it.

**Comment by alexey-pelykh** [comment]
@abhinavsingh your turn to tame your CI :)

**Comment by abhinavsingh** [comment]
> 

I restarted the failed jobs.   Due to certain integration tests, which uses upstream httpbin.org and pythonhosted.org, we end up receiving server side / rate limiting errors.   As a result, these tests can fail sometimes.   Usually, restarting them always makes them work.

**Comment by abhinavsingh** [comment]
@alexey-pelykh So workflow went fine after restarting failed jobs.   Docker and GHCR are related to secrets and permissions for PRs I suppose.   I'll take another stab at it sometime tonight or morrow and merge it.   Thank you for adding this capability.  It certainly can be used to scale proxy.py even further on a single host machine.

**Comment by alexey-pelykh** [comment]
@abhinavsingh how's it going on your end? any changes needed?

**Comment by abhinavsingh** [comment]
> @abhinavsingh how's it going on your end? any changes needed?

Hi @alexey-pelykh , apologies, I got derailed a release work past/this week.   Promise, I'll take a look again and merge this weekend.   Thank you.

**Comment by abhinavsingh** [comment]
@alexey-pelykh LGTM, made few minor edits directly in the PR.   Workflow revealed failures which led me to  https://github.com/home-assistant/core/issues/91283 🤦‍♂️ Damn.   Will need to fix the `develop` branch workflow before merging this PR

**Pull request merged**